llvm/lib/Target/Mips/MipsISelLowering.cpp

```diff
@@ -1129,9 +1147,20 @@ static SDValue attemptOrToIns(SDValue &And, SDValue &Right, SDNode *N,
     if (!(CN = dyn_cast<ConstantSDNode>(Right.getOperand(1))))
       return SDValue();
     unsigned ShiftAmount = CN->getZExtValue();
-    // Second check makes sure that all upper bits are picked up.
-    if ((ShiftAmount != SMPos0) || (SMPos0 + SMSize0 != TSize))
+
+    if ((ShiftAmount != SMPos0))
       return SDValue();
+
+    if ((SMPos0 + SMSize0 != TSize)) {
+      // If a register holds a value that is zero extended
+      // and if the actual size of that value can fit in SMSize0
+      // then we will allow replacement with the INS instruction
+      // because we are sure that we have only leading zeros
+      KnownBits Known = DAG.computeKnownBits(Right.getOperand(0));
+      if (Known.countMaxPopulation() > SMSize0)
```

Comment:
This seems wrong. countMaxPopulation() returns the count of the number of bits that could be one, the maximum possible value of `popcnt()`, without saying anything about *where* those 1 bits are. Don't we need to make sure here that they're only in the lowest SMSize0 bits?

Reply:
Yes, I agree, we don't have that guarantee with the current solution. Maybe it is enough to check if ```Known.getMaxValue().getActiveBits()``` is smaller than ```SMSize0``` and to allow ```INS``` to be generated in that case? I think that should give us maximum number of bits set to 1 starting from the lowest positions.